pkg/networkmanager/firewall.jsx

```diff
@@ -0,0 +1,265 @@
+    }
+
+    var columns = [
+        { name: props.service.name, header: true },
+        <div>
+            { tcp.map(p => p.port).join(', ') }
+        </div>,
+        <div>
+            { udp.map(p => p.port).join(', ') }
+        </div>,
+        <button className="btn btn-danger pficon pficon-delete"
+                onClick={onRemoveService}/>
+    ];
+
+    var tabs = [
+        { name: _("Details"), renderer: () => <p>{props.service.description}</p> }
+    ];
+
+    return <ListingRow key={props.service.id}
+                       rowId={props.service.id}
+                       columns={columns}
+                       tabRenderers={tabs}/>;
+}
+
+class AddServicesDialogBody extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            selected: new Set()
+        };
+
+        this.onToggleService = this.onToggleService.bind(this);
+    }
+
+    componentDidMount() {
+        firewall.getAvailableServices()
+            .then(services => this.setState({ services: services }));
+    }
+
+    onToggleService(event) {
+        var service = event.target.id;
+        var enabled = event.target.checked;
+
+        this.setState(oldState => {
+            var selected = new Set(oldState.selected);
+
+            if (enabled)
+                selected.add(service)
+            else
+                selected.delete(service);
+
+            this.props.selectionChanged(selected);
+
+            return {
+                selected: selected
+            };
+        });
+    }
+
+    render() {
+        if (!this.state.services) {
+            return (
+                <div className="modal-body">
+                    <div className="spinner spinner-lg" />
+                </div>
+            );
+        }
+
+        return (
+            <div id="add-services-dialog" className="modal-body">
+                <ul className="list-group dialog-list-ct">
+                    {
+                        this.state.services.map(s => (
+                            <li className="list-group-item">
+                                <label>
+                                    <input id={s.id}
+                                           type="checkbox"
+                                           checked={this.state.selected.has(s.id)}
+                                           onClick={this.onToggleService} />
+                                    &nbsp;
+                                    <span>{s.name}</span>
+                                </label>
+                            </li>
+                        ))
+                    }
+                </ul>
+            </div>
+        );
+    }
+}
+
+export class Firewall extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            installed: firewall.installed,
+            enabled: firewall.enabled,
+            services: firewall.services,
+            enabledServices: firewall.enabledServices
+        };
```

Comment:
It feels redundant to keep a verbatim copy of the global `firewall` state here and in `onFirewallChanged`. Isn't it enough to just call `setState({ firewall })` in `onFirewallChanged`, and initialize it the same way here? Or am I missing something?

Reply:
I'd still like to discuss the above.